src/backend/distributed/utils/colocation_utils.c

```diff
@@ -14,14 +14,322 @@
+												   sourceDistributionColumnType);
+		UpdateColocationGroup(sourceRelationId, sourceColocationId);
+	}
+
+	/* finally set colocation group for the target relation */
+	UpdateColocationGroup(targetRelationId, sourceColocationId);
+
+	heap_close(pgDistColocation, NoLock);
+}
+
+
+/*
+ * ErrorIfShardPlacementsNotColocated checks if the shard placements of the
+ * given two relations are physically colocated. If they are not colocated,
+ * it errors out.
+ *
+ * Note that, this functions assumes that both tables are hash distributed.
+ */
+static void
+ErrorIfShardPlacementsNotColocated(Oid leftRelationId, Oid rightRelationId)
+{
+	List *leftShardIntervalList = NIL;
+	List *rightShardIntervalList = NIL;
+	ListCell *leftShardIntervalCell = NULL;
+	ListCell *rightShardIntervalCell = NULL;
+	char *leftRelationName = NULL;
+	char *rightRelationName = NULL;
+	uint32 leftShardCount = 0;
+	uint32 rightShardCount = 0;
+
+	/* get sorted shard interval lists for both tables */
+	leftShardIntervalList = LoadShardIntervalList(leftRelationId);
+	rightShardIntervalList = LoadShardIntervalList(rightRelationId);
+
+	/* prevent concurrent placement changes */
+	LockShardListMetadata(leftShardIntervalList, ShareLock);
+	LockShardListMetadata(rightShardIntervalList, ShareLock);
+
+	leftRelationName = get_rel_name(leftRelationId);
+	rightRelationName = get_rel_name(rightRelationId);
+
+	leftShardCount = list_length(leftShardIntervalList);
+	rightShardCount = list_length(rightShardIntervalList);
+
+	if (leftShardCount != rightShardCount)
+	{
+		ereport(ERROR, (errmsg("shard counts don't match for %s and %s",
+							   leftRelationName, rightRelationName)));
+	}
+
+	/* compare shard intervals one by one */
+	forboth(leftShardIntervalCell, leftShardIntervalList,
+			rightShardIntervalCell, rightShardIntervalList)
+	{
+		ShardInterval *leftInterval = (ShardInterval *) lfirst(leftShardIntervalCell);
+		ShardInterval *rightInterval = (ShardInterval *) lfirst(rightShardIntervalCell);
+
+		List *leftPlacementList = NIL;
+		List *rightPlacementList = NIL;
+		List *sortedLeftPlacementList = NIL;
+		List *sortedRightPlacementList = NIL;
+		ListCell *leftPlacementCell = NULL;
+		ListCell *rightPlacementCell = NULL;
+
+		uint64 leftShardId = leftInterval->shardId;
+		uint64 rightShardId = rightInterval->shardId;
+
+		bool shardsIntervalsEqual = ShardsIntervalsEqual(leftInterval, rightInterval);
+		if (!shardsIntervalsEqual)
+		{
+			ereport(ERROR, (errmsg("shard intervals don't match for %s and %s",
+								   leftRelationName, rightRelationName)));
+		}
+
+		leftPlacementList = ShardPlacementList(leftShardId);
+		rightPlacementList = ShardPlacementList(rightShardId);
+
+		/* sort shard placements according to the node */
+		sortedLeftPlacementList = SortList(leftPlacementList,
+										   CompareShardPlacementsByNode);
+		sortedRightPlacementList = SortList(rightPlacementList,
+											CompareShardPlacementsByNode);
+
+		/* compare shard placements one by one */
```

Comment:
To simplify things and remove some code, I'm thinking of the following. Could you also share your thoughts on this?

```
leftPlacementList = ShardPlacementList(leftShardId);
rightPlacementList = ShardPlacementList(rightShardId);

-- move this upper level in the function right after shard count check
if (TableShardReplicationFactor(leftTable) != TableShardReplicationFactor(rightTable))
error (replication factors not equal and some more detail)

/* I've typo below */
leftFinializedPlacementList = FinalizedShardPlacementList(leftShardId);
rightFinializedPlacementList = FinalizedShardPlacementList(leftShardId);

If (list_length(leftPlacementList) != list_length(leftFinializedPlacementList) )
error (Table has non-finialized placement)

If (list_length(rightPlacementList) != list_length(rightFinializedPlacementList) )
error (Table has non-finialized placement)



intersectedList = IntersectPlacementList(leftFinializedPlacementList, rightPlacementList);

if list_length(intersectedList) != leftFinializedPlacementList
ereport(ERROR, (errmsg("some shard placements are not colocated for %s "
 "and %s", leftRelationName, rightRelationName)));

```

Now that we error out for any placements which are not healthy.

I'm really not worried about the performance of this given that it's a one time operation and I don't expect this to run >1 sec even for a table pair with 1024 shards